File: scratch/growth/tmbModel/increment/growth.R



# path_current <- Sys.getenv('PATH')
# path_new <- paste0('c:\\Rtools\\bin;c:\\Rtools\\mingw_32\\bin;', 
#                    path_current)
# Sys.setenv(PATH=path_new)

# load the TMB package
require(TMB)


# compile the c++ file and make available to R
compile("scratch/growth/tmbModel/growth.cpp")
dyn.load(dynlib("scratch/growth/tmbModel/growth"))


# read in the data
load('data/data_processed/trawlBiol.Rdata') # biol
tdc <- subset(biol, COMNAME == 'ATLANTIC COD')

yss <- substr(tdc$CRUISE, start=1, stop=4)
tdc$YEAR <- as.numeric(yss)
tdcsub0 <- tdc[,c('YEAR', 'AGE', 'LENGTH')]
tdcsub1 <- subset(tdcsub0, AGE > 0)
tdcsub <- tdcsub1[complete.cases(tdcsub1),]


# number of years to include in the analysis will be the
# number of years there are altogether plus an initialization
# period of mxage
mxage <- max(tdcsub$AGE)
mxyear <- max(tdcsub$YEAR)
nyear <- length(unique(tdcsub$YEAR))
nage <- length(unique(tdcsub$AGE))

data <- list(nyear = nyear,
             nage = nage,
             mxage = mxage,
             mxyear = mxyear,
             nrec = nrow(tdcsub),
             YEAR = tdcsub$YEAR,
             AGE = tdcsub$AGE,
             LENGTH = tdcsub$LENGTH)

parameters <- list(log_Linf = log(150), 
                   log_K = log(0.2),
                   log_oe_L = 0.5)

# Set parameter bounds
lb <- c(log_Linf = log(80),
        log_K = log(0.001),
        log_oe_L = log(0.001))

ub <- c(log_Linf = log(200),
        log_K = log(1),
        log_oe_L = log(2))


map_par <- list(
  # log_Linf = factor(NA),
  # log_K = factor(NA)
  # log_oe_L = factor(NA)
)


# map_par <- list() # all parameters active


# make an objective function
obj <- MakeADFun(data = data,
                 parameters = parameters,
                 DLL = "growth",
                 map = map_par
)


# index for active parameter bounds
# (note that R does not give an error message if the wrong
# number of parameter bounds are given
# if having estimation problems parameter bounds are a good
# place to check)
active_idx <- !names(lb) %in% names(map_par)



# run the model using the R function nlminb()
opt <- nlminb(start=obj$par, 
              objective=obj$fn, 
              gradient=obj$gr, 
              lower=unlist(lb[active_idx]),
              upper=unlist(ub[active_idx]))



# estimates and standard errors (run after model optimization)
sdrep <- sdreport(obj)


# list of variables exported using REPORT() function in c++ code
rep <- obj$report()

plot(LENGTH ~ AGE, data=tdcsub, ylim=c(0, 155),
     xlim=c(0, 20))
vb <- function(x) rep$Linf * (1-exp(-rep$K*x))
curve(vb, from=0, to=20, add=TRUE, col='blue', lwd=5)


m <- nls(LENGTH~Linf * (1-exp(-K*(AGE-t0))),
         data=tdcsub,
         start=list(Linf=120, K=0.2, t0=0))
vb2 <- function(x) coef(m)['Linf'] * 
                   (1-exp(-coef(m)['K']*(x-coef(m)['t0'])))
curve(vb2, from=0, to=20, add=TRUE, col='red', lwd=2)


m2 <- nls(LENGTH~Linf * (1-exp(-K*(AGE))),
         data=tdcsub,
         start=list(Linf=120, K=0.2))
vb3 <- function(x) coef(m2)['Linf'] * 
  (1-exp(-coef(m2)['K']*x))
curve(vb3, from=0, to=20, add=TRUE, col='green', lwd=3)

File: scratch/growth/tmbModel/increment/growth.cpp




#include <TMB.hpp>  // Links in the TMB libraries

// must include the next two lines
template<class Type>
Type objective_function<Type>::operator() ()
{
  
  // DATA

  // index bounds
  DATA_INTEGER(nyear);
  DATA_INTEGER(nage);
  DATA_INTEGER(mxage);
  DATA_INTEGER(mxyear);
  DATA_INTEGER(nrec);

  // vectors of observation data
  DATA_IVECTOR(YEAR);
  DATA_IVECTOR(AGE);
  DATA_VECTOR(LENGTH);


  // Parameters
  PARAMETER(log_Linf);
    Type Linf = exp(log_Linf);
  PARAMETER(log_K);
    Type K = exp(log_K);
  PARAMETER(log_oe_L);
    Type oe_L = exp(log_oe_L);


  // Containers
  matrix<Type> L(nyear+nage,nage+1);
  vector<Type> Lrec(nrec);
  int indy;
  int inda;
  Type predL;
  Type NLL = 0;
  
 
 
  // Model

  // predictions for length@age in every year of the model

  for(int i=0; i<L.cols(); i++){
    L(0,i) = 0;
  }
  for(int i=0; i<L.rows(); i++){
    L(i,0) = 0;
  }
  // fill in initial length estimates
  for(int y=1; y<nyear+nage; y++){
    for(int a=1; a<nage+1; a++){
      L(y,a) = L(y-1,a-1) + (Linf - L(y-1,a-1)) * (1-exp(-K));
    }
  }


  // negative log likelihood
  for(int i=0; i<nrec; i++){
    //double check the year idx is right ... age was wrong and 
    // I removed the +1.
    indy = nyear-(mxyear-YEAR(i)+1) + nage;
    inda = nage-(mxage-AGE(i));
    predL = L(indy,inda);
    Lrec(i) = predL;
    NLL -= dnorm(log(LENGTH(i)), log(predL), oe_L, true);
  }


  // Report

  REPORT(NLL);
  REPORT(L);
  REPORT(Linf);
  REPORT(K);
  REPORT(oe_L);
  REPORT(Lrec);

  
  return(NLL);
  
}
    

